dtfs-central-api/src/v1/controllers/utilisation-report-service/post-report-data-validation.controller/index.ts

```diff
@@ -0,0 +1,30 @@
+import { ApiError, CustomExpressRequest } from '@ukef/dtfs2-common';
+import { Response } from 'express';
+import { HttpStatusCode } from 'axios';
+import { validateUtilisationReportCsvData } from '../../../../services/utilisation-report-data-validator';
+import { PostReportDataValidationPayload } from '../../../routes/middleware/payload-validation/validate-post-validate-utilisation-report-data-payload';
+
+export type PostReportDataValidationRequest = CustomExpressRequest<{
+  reqBody: PostReportDataValidationPayload;
+}>;
+
+/**
+ * Controller to handle post request to report data validation route
+ * @param req - The request
+ * @param res - The response
+ */
+export const postReportDataValidation = (req: PostReportDataValidationRequest, res: Response) => {
+  const { reportData } = req.body;
+
+  try {
+    const csvValidationErrors = validateUtilisationReportCsvData(reportData);
+    return res.status(HttpStatusCode.Ok).send({ csvValidationErrors });
+  } catch (error) {
+    const errorMessage = 'Failed to validate report data';
+    console.error(errorMessage, error);
+    if (error instanceof ApiError) {
+      return res.status(error.status).send(`${errorMessage}: ${error.message}`);
+    }
+    return res.status(HttpStatusCode.InternalServerError).send(errorMessage);
```

Comment:
I think this is the standard pattern we have in our controllers for handling `ApiError`s, so keen to keep it for consistency. Note that the status used and message is slightly different between lines 26 and 28 so it isn't the same.

Reply:
Thanks @RLavender98 - looks like this is only in place for some utilisation report and some amendment controllers though so perhaps it isn't a standard pattern.

I did notice the statuses are different - what's the reasoning for this, why do we need to do this? Do any consumers of this endpoint react differently depending on the status?

Reply:
My understanding is ApiError is a new pattern that was brought in shortly before I joined the project which is the reason it isn't used in more places. 

The idea is we can have standard errors that extend this type and that store the status on themselves so that the controller error handling can be simpler.

Say you have an endpoint that could throw 10 different kinds of expected errors. We could check for each of these errors here one by one to allow us to return the correct status code and message depending on the error type (some might be errors that need to be BadRequest, or Forbidden for example). Instead, using the ApiError class, we have each of those 10 errors extend the ApiError type and then we can handle them all in one nice block since we get the error to own it's own status. e.g. `InvalidPayloadError` extends `ApiError` and it's error code is a `BadRequest`.

The line on 28 is for any unexpected errors, in which case that would fall under InternalServerError and we want to send a more generic message since we don't know what is in the error message since we didn't write it ourselves and we don't want to accidentally reveal sensitive information which could be in the error message.

As for consumers responding differently, I don't think they do, but I don't think that is a good reason to not use correct error codes. Distinct error codes can also be useful for logging and debugging purposes. Also if we ever wanted to have alerts set up on different status codes for example, this would be really useful, because then we could track 500s and 400s separately which would be pretty nice.